sched/semaphore/spinlock.c

```diff
@@ -384,4 +388,138 @@ void spin_clrbit(FAR volatile cpu_set_t *set, unsigned int cpu,
 }
 #endif
 
+#if defined(CONFIG_RW_SPINLOCK)
+
+/****************************************************************************
+ * Name: rw_read_spin_lock
+ *
+ * Description:
+ *   If this CPU does not already hold the spinlock, then loop until the
+ *   spinlock is successfully locked.
+ *
+ *   This implementation is non-reentrant and set a bit of lock.
+ *
+ *  The priority of reader is higher than writter if a reader hold the
+ *  lock, a new reader can get its lock but writer can't get this lock.
+ *
+ * Input Parameters:
+ *   lock - A reference to the spinlock object to lock.
+ *
+ * Returned Value:
+ *   None.  When the function returns, the spinlock was successfully locked
+ *   by this CPU.
+ *
+ * Assumptions:
+ *   Not running at the interrupt level.
+ *
+ ****************************************************************************/
+
+void rw_read_spin_lock(FAR volatile rw_spinlock_t *lock)
+{
+  while (atomic_load(lock) == -1)
+    {
+      SP_DSB();
+      SP_WFE();
+    }
+
+  atomic_fetch_add(lock, 1);
```

Comment:
race since another writer may hold lock after line 419 and before line 425

Reply:
Yes. You're right.
I need to think how to implement this again.

Reply:
I already modify the implementation.

Reply:
Fixed.